src/EventHandler.ts

```diff
@@ -0,0 +1,248 @@
+            this.pendingTouchEventStart = null;
+            this.pendingTouchEventEnd = event;
+            setTimeout(this.handleTap, 200);
+            return;
+        }
+
+        this.pendingTouchEventEnd = null;
+
+        if (event.timeStamp - this.pendingTouchEventStart.timeStamp > 500) {
+            // This is a slow swipe / highlight. Let the browser handle it.
+            this.pendingTouchEventStart = null;
+            return;
+        }
+
+        // This is a swipe. 
+        const slope = (startTouch.clientY - endTouch.clientY) / (startTouch.clientX - endTouch.clientX);
+        if (Math.abs(slope) > 0.5) {
+            // This is a mostly vertical swipe. Ignore.
+            this.pendingTouchEventStart = null;
+            return;
+        }
+
+        // This was a horizontal swipe.
+        if (relativeXDistance < 0) {
+            this.onBackwardSwipe(event);
+        } else {
+            this.onForwardSwipe(event);
+        }
+        this.pendingTouchEventStart = null;
+    }
+
+    private handleClick = (): void => {
+        if (!this.pendingMouseEventEnd) {
+            // Another click happened already.
+            return;
+        }
+
+        if (this.checkForLink(this.pendingMouseEventEnd)) {
+            // This was a single click on a link. Do nothing.
+            this.pendingMouseEventEnd = null;
+            return;
+        }
+
+        // This was a single click.
+        const x = this.pendingMouseEventEnd.clientX;
+        const width = window.innerWidth;
+        if (x / width < 0.3) {
+            this.onLeftTap(this.pendingMouseEventEnd);
+        } else if (x / width > 0.7) {
+            this.onRightTap(this.pendingMouseEventEnd);
+        } else {
+            this.onMiddleTap(this.pendingMouseEventEnd);
+        }
+        this.pendingMouseEventEnd = null;
+        return;
+    }
+
+    private handleTap = (): void => {
+        if (!this.pendingTouchEventEnd) {
+            // Another tap happened already.
+            return;
+        }
+
+        if (this.checkForLink(this.pendingTouchEventEnd)) {
+            // This was a single tap on a link. Do nothing.
+            this.pendingTouchEventEnd = null;
+            return;
+        }
+
+        // This was a single tap.
+        const touch = this.pendingTouchEventEnd.changedTouches[0];
+        if (!touch) {
+            // Somehow we got a touch event with no touches.
+            return;
+        }
+
+        const x = touch.clientX;
+        const width = window.innerWidth;
+        if (x / width < 0.3) {
+            this.onLeftTap(this.pendingTouchEventEnd);
+        } else if (x / width > 0.7) {
+            this.onRightTap(this.pendingTouchEventEnd);
+        } else {
+            this.onMiddleTap(this.pendingTouchEventEnd);
+        }
+        this.pendingTouchEventEnd = null;
+        return;
+    }
+
+    private checkForLink = (event: MouseEvent | TouchEvent): boolean => {
+        let nextElement = event.target as Element;
```

Comment:
Since `event.target` can also be a document or window, at least in general, it would be better to write the code like this:

``` typescript
let target = event.target
if(target instanceof Element) {
  // In this block, TypeScript will know `target` has type `Element` and no
  // unsafe cast is needed.
  ...
} else {
  return false;
}
```

If you haven't seen this construct before, you may want to look at [this article](https://basarat.gitbooks.io/typescript/docs/types/typeGuard.html).